test/local_test.go

```diff
@@ -0,0 +1,69 @@
+package test
+
+import (
+	"errors"
+	"gopkg.in/yaml.v3"
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"testing"
+)
+
+type parsedCRD struct {
+	APIVersion string `yaml:"apiVersion"`
+	Kind       string `yaml:"kind"`
+	Metadata   struct {
+		Annotations struct {
+			EksAmazonawsComRoleArn              string `yaml:"eks.amazonaws.com/role-arn"`
+			EksAmazonawsComStsRegionalEndpoints string `yaml:"eks.amazonaws.com/sts-regional-endpoints"`
+		} `yaml:"annotations"`
+		Name string `yaml:"name"`
+	} `yaml:"metadata"`
+}
+
+func TestServiceAccountsUseStsRegionalEndpoint(t *testing.T) {
+	addonsPath := "../modules/cluster/addons/"
+	addonFiles, err := ioutil.ReadDir(addonsPath)
+	if err != nil {
+		t.Error(err)
+	}
+	for _, v := range addonFiles {
+		fileName := v.Name()
+		if filepath.Ext(fileName) == ".yaml" {
+			crds := readManifest(addonsPath + fileName)
+			for _, v := range crds {
+				if v.Kind == "ServiceAccount" {
+					if v.Metadata.Annotations.EksAmazonawsComRoleArn != "" && v.Metadata.Annotations.EksAmazonawsComStsRegionalEndpoints != "true" {
+						t.Errorf("Service account %s role arn %s has EksAmazonawsComStsRegionalEndpoints=%s, should be 'true'", v.Metadata.Name, v.Metadata.Annotations.EksAmazonawsComRoleArn, v.Metadata.Annotations.EksAmazonawsComStsRegionalEndpoints)
+
+					}
+				}
+			}
+		}
+	}
+}
+
+func readManifest(filename string) []parsedCRD {
+	f, err := os.Open(filename)
+	if err != nil {
+		panic(err)
+	}
+	d := yaml.NewDecoder(f)
+	var crds []parsedCRD
+	for {
+		var crd parsedCRD
+		err := d.Decode(&crd)
+		if &crd == nil {
+			continue
+		}
+		if errors.Is(err, io.EOF) {
```

Comment:
Why don't you have general `if err != nil {}`? This can capture `io.EOF` but cannot catch other errors.

Reply:
Hm, I think the implementation here is trying to parse every yaml entity as a crd, so errors are expected for the not-crds.

That said, would this condition ever fire? I think the `== nil { continue }` would win.

Reply:
this condition fires when all elements in the yaml file have been parsed, so if we remove this we get an infinite loop

I think this will cover the points?

```
		if errors.Is(err, io.EOF) {
			break
		}
		if err != nil {
			return nil, err
		}
```

Reply:
Why doesn't `if &crd == nil { continue }` win? If we've reached the end of the file, I'd expect that we would not got a crd populated.

Reply:
Oh! Because it's *never* nil, because it's a defined variable which you're taking a reference from.

Reply:
So maybe...
``
if errors.Is(err, io.EOF) {
    break
}

if crd == parsedCRD{} {
   continue
}
```

I think your `if err != nil { return nil, err }` would not work. Because I *think* you'd get an error if the given yaml token couldn't be unmarshalled into the parsedCRD struct.